File: packages/python/keyring/starfish_keyring/recipients.py
"""Collection-scoped recipient management helpers.

These wrap the low-level keyring functions in :mod:`starfish_sdk.keyring` with
HTTP-aware I/O via :class:`StarfishClient`. The keyring document for a
collection lives at the conventional path ``<collection>/_keyring`` and is
fetched/pushed using the ``/pull/`` and ``/push/`` route prefixes.

Hash-based conflict detection is preserved: each push uses the hash from the
prior pull as ``base_hash``. Callers may retry on :class:`ConflictError`.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, TypedDict

from starfish_sdk.types import StarfishHttpError

from .keyring import (
    Keyring,
    add_recipient as keyring_add_recipient,
    rotate_epoch,
    unwrap_from_entry,
    verify_entry_signature,
)

if TYPE_CHECKING:
    from starfish_sdk.client import StarfishClient


def keyring_path_for(collection_name: str) -> str:
    """Document path used to store a collection's keyring.

    Returns ``<collection>/_keyring``. The helpers below prefix this with
    ``/pull/`` or ``/push/`` as required by the route layer.
    """
    return f"{collection_name}/_keyring"


class RecipientRef(TypedDict, total=False):
    """A recipient referenced by its KEM (X25519) public key, with metadata."""

    subKem: str
    userId: str
    label: str


class AdderKeys(TypedDict):
    """Adder's keypair material — device must already be in the current epoch."""

    edPriv: str
    edPub: str
    kemPriv: str


class ListedRecipient(TypedDict):
    """One recipient projected for listing."""

    subKem: str
    addedBy: str
    addedAt: int


def _pull_path_for(collection_name: str) -> str:
    return f"/pull/{keyring_path_for(collection_name)}"


def _push_path_for(collection_name: str) -> str:
    return f"/push/{keyring_path_for(collection_name)}"


async def _pull_keyring(
    client: StarfishClient, collection_name: str
) -> tuple[Keyring, str] | None:
    """Pull the keyring; ``None`` if the document doesn't exist yet (HTTP 404)."""
    try:
        result = await client.pull(_pull_path_for(collection_name))
    except StarfishHttpError as exc:
        if exc.status == 404:
            return None
        raise
    # result is a PullResult — data is a dict suitable for Keyring.from_dict.
    data: dict[str, Any] = result.data  # type: ignore[union-attr]
    return Keyring.from_dict(data), result.hash  # type: ignore[union-attr]


def _recover_current_cek(
    keyring: Keyring,
    adder_kem_priv_hex: str,
    trusted_adders: set[str] | None = None,
) -> bytes:
    """Locate the adder's entry in the current epoch and recover the CEK.

    Probes each entry until one decrypts successfully — this lets callers pass
    just the KEM private key without also having to identify which ``subKem``
    is theirs.

    Each candidate entry's ``addedSig`` is verified before the unwrap
    attempt: tampered audit signatures (e.g. server-mutated ``addedBy``)
    cause the entry to be skipped (with a logged warning), so the adder
    never trusts unattested wrap material and a single corrupted entry
    does not prevent recovery via a sibling entry in the same epoch.

    When ``trusted_adders`` is provided, entries whose ``added_by`` is not in
    the set are skipped — the ``addedSig`` is self-attesting, so a hostile
    server can REPLACE the adder's own entry with one wrapping an
    attacker-chosen CEK to the adder's KEM pubkey and self-sign it; pinning the
    trusted adders closes that substitution.
    """
    import logging

    _log = logging.getLogger(__name__)
    epoch_key = str(keyring.current_epoch)
    epoch = keyring.epochs.get(epoch_key)
    if epoch is None:
        raise ValueError(f"Epoch {keyring.current_epoch} not found in keyring")

    # A valid epoch has unique subKems (enforced on write by add_recipient).
    # Duplicates mean the keyring was tampered with — e.g. a hostile server
    # injected an entry wrapping an attacker-chosen CEK to this adder's key.
    # Fail closed rather than risk recovering and re-wrapping a forged CEK.
    seen_sub_kems: set[str] = set()
    for entry in epoch.wrapped_keys:
        if entry.sub_kem in seen_sub_kems:
            raise ValueError(
                f"Keyring epoch {keyring.current_epoch} has duplicate entries "
                f"for subKem={entry.sub_kem} (tampering)"
            )
        seen_sub_kems.add(entry.sub_kem)

    last_err: Exception | None = None
    for entry in epoch.wrapped_keys:
        if trusted_adders is not None and entry.added_by not in trusted_adders:
            _log.warning(
                "skipping entry subKem=%s in epoch %d: addedBy %s is not a trusted adder",
                entry.sub_kem,
                keyring.current_epoch,
                entry.added_by,
            )
            continue
        if not verify_entry_signature(entry, keyring.current_epoch):
            _log.warning(
                "skipping entry subKem=%s in epoch %d: addedSig verification failed",
                entry.sub_kem,
                keyring.current_epoch,
            )
            continue
        try:
            return unwrap_from_entry(entry, adder_kem_priv_hex)
        except Exception as err:  # noqa: BLE001 — try every entry
            last_err = err
    raise ValueError(
        f"Adder has no usable entry in current epoch {keyring.current_epoch}"
    ) from last_err


def _require_trusted_adders(trusted_adders: list[str] | None, fn: str) -> set[str]:
    """Resolve the mandatory ``trusted_adders`` pin into a set, or raise.

    The mutation helpers recover the current CEK from a server-supplied keyring;
    without a provenance pin a hostile server could substitute a forged entry
    (the ``addedSig`` is self-attesting). Fail closed rather than mutate off
    unverified key material.
    """
    if not trusted_adders:
        raise ValueError(
            f"{fn}: `trusted_adders` is required — pass the Ed25519 pubkey(s) you trust to "
            f"grant keyring access (e.g. the collection owner's root key). Without it a hostile "
            f"server could substitute a wrapped-key entry (the addedSig is self-attesting)."
        )
    return set(trusted_adders)


async def add_recipient(
    client: StarfishClient,
    collection_name: str,
    recipient: RecipientRef,
    adder: AdderKeys,
    *,
    trusted_adders: list[str] | None = None,
) -> None:
    """Add a new recipient to the current epoch (no rotation).

    Pulls the current keyring, unwraps the CEK using the adder's KEM private
    key, appends a new entry signed by the adder, and pushes the updated
    keyring back using the prior hash as ``base_hash``.

    ``trusted_adders`` is the same pin ``create_keyring_encryptor`` accepts.
    The ``addedSig`` is self-attesting, so a hostile server can REPLACE the
    adder's own entry with one wrapping an attacker-chosen CEK to the adder's
    (public) KEM key and self-sign it; ``_recover_current_cek`` would then
    unwrap that forged CEK and re-wrap it for the new recipient. Pinning the
    trusted adders ignores such entries. Strongly recommended whenever the
    server is not fully trusted.

    Raises ``ValueError`` if the keyring document does not exist yet.
    """
    pulled = await _pull_keyring(client, collection_name)
    if pulled is None:
        raise ValueError(
            f"Cannot add recipient: no keyring exists at {keyring_path_for(collection_name)}. "
            f"Create the keyring first."
        )
    keyring, prev_hash = pulled

    trusted = _require_trusted_adders(trusted_adders, "add_recipient")
    current_cek = _recover_current_cek(keyring, adder["kemPriv"], trusted)
    next_keyring = keyring_add_recipient(
        keyring,
        adder_ed_priv_hex=adder["edPriv"],
        adder_ed_pub_hex=adder["edPub"],
        current_cek=current_cek,
        recipient_kem_hex=recipient["subKem"],
    )

    await client.push(
        _push_path_for(collection_name),
        next_keyring.to_dict(),
        prev_hash,
    )


async def remove_recipient(
    client: StarfishClient,
    collection_name: str,
    remove_sub_kems: list[str],
    adder: AdderKeys,
    *,
    trusted_adders: list[str] | None = None,
) -> dict[str, int]:
    """Rotate the epoch, dropping the named recipients.

    When ``trusted_adders`` is set, entries written by an untrusted adder (e.g.
    a recipient a hostile server injected) are not carried into the new epoch —
    the rotation would otherwise re-wrap the fresh CEK for them.

    Returns ``{"newEpoch": <int>}``. Raises ``ValueError`` if the keyring
    document does not exist yet.
    """
    pulled = await _pull_keyring(client, collection_name)
    if pulled is None:
        raise ValueError(
            f"Cannot remove recipient: no keyring exists at {keyring_path_for(collection_name)}."
        )
    keyring, prev_hash = pulled

    epoch_key = str(keyring.current_epoch)
    epoch = keyring.epochs.get(epoch_key)
    if epoch is None:
        raise ValueError(f"Epoch {keyring.current_epoch} not found in keyring")

    trusted = _require_trusted_adders(trusted_adders, "remove_recipient")
    remove_set = set(remove_sub_kems)
    retained = [
        e.sub_kem
        for e in epoch.wrapped_keys
        if e.sub_kem not in remove_set and e.added_by in trusted
    ]

    rotated, _new_cek = rotate_epoch(
        keyring,
        adder_ed_priv_hex=adder["edPriv"],
        adder_ed_pub_hex=adder["edPub"],
        retained_recipients=retained,
    )

    await client.push(
        _push_path_for(collection_name),
        rotated.to_dict(),
        prev_hash,
    )

    return {"newEpoch": rotated.current_epoch}


async def list_recipients(
    client: StarfishClient,
    collection_name: str,
    *,
    trusted_adders: list[str] | None = None,
) -> dict[str, Any]:
    """List recipients in the current epoch, filtered by provenance.

    Returns ``{"epoch": int, "recipients": [{"subKem", "addedBy", "addedAt"}, ...]}``.
    Only entries whose ``addedBy`` is in ``trusted_adders`` AND whose ``addedSig``
    verifies are returned — mirroring :func:`create_keyring_encryptor`, so a forged
    or server-substituted entry never surfaces in a membership/admin view.

    ``trusted_adders`` is REQUIRED (fail-closed): the keyring is fetched from an
    untrusted server and the per-entry ``addedSig`` is self-attesting, so without a
    provenance pin a hostile server could spoof the listing. Pass the Ed25519
    pubkey(s) you trust to grant access (e.g. the collection owner's root key).
    If the keyring document doesn't exist yet, returns ``{"epoch": 0, "recipients": []}``.
    """
    trusted = _require_trusted_adders(trusted_adders, "list_recipients")
    pulled = await _pull_keyring(client, collection_name)
    if pulled is None:
        return {"epoch": 0, "recipients": []}
    keyring, _ = pulled

    epoch = keyring.epochs.get(str(keyring.current_epoch))
    if epoch is None:
        return {"epoch": keyring.current_epoch, "recipients": []}

    recipients: list[ListedRecipient] = [
        {"subKem": e.sub_kem, "addedBy": e.added_by, "addedAt": e.added_at}
        for e in epoch.wrapped_keys
        if e.added_by in trusted and verify_entry_signature(e, keyring.current_epoch)
    ]
    return {"epoch": keyring.current_epoch, "recipients": recipients}


async def current_epoch(client: StarfishClient, collection_name: str) -> int:
    """Return the current epoch number; ``0`` if no keyring exists yet."""
    pulled = await _pull_keyring(client, collection_name)
    if pulled is None:
        return 0
    keyring, _ = pulled
    return keyring.current_epoch

File: packages/ts/keyring/src/recipients.ts
/**
 * Collection-scoped recipient management helpers.
 *
 * These wrap the low-level keyring functions in `keyring.ts` with HTTP-aware
 * I/O via `StarfishClient`. The keyring document for a collection lives at the
 * conventional path `<collection>/_keyring` and is fetched/pushed using the
 * `/pull/` and `/push/` route prefixes.
 *
 * Hash-based conflict detection is preserved: each push uses the hash from
 * the prior pull as `baseHash`. Callers may retry on `ConflictError`.
 */

import type { Alg } from "@drakkar.software/starfish-protocol"
import type { StarfishClient } from "@drakkar.software/starfish-client"
import { StarfishHttpError } from "@drakkar.software/starfish-client"
import type { Keyring, WrappedKeyEntry } from "./keyring.js"
import {
  addRecipient as keyringAddRecipient,
  rotateEpoch,
  unwrapFromEntry,
  verifyEntrySignature,
} from "./keyring.js"

/**
 * Computes the document path used to store a collection's keyring.
 * Returns `${collectionName}/_keyring`. The helpers below prefix this with
 * `/pull/` or `/push/` as required by the route layer.
 */
export function keyringPathFor(collectionName: string): string {
  return `${collectionName}/_keyring`
}

/** A recipient referenced by its KEM public key, with optional metadata. */
export interface RecipientRef {
  /** Recipient KEM pubkey (hex) of suite `kemAlg` (X25519 for ed25519). */
  subKem: string
  /** Recipient KEM suite. Absent ⇒ `ed25519` (X25519). */
  kemAlg?: Alg
  userId?: string
  label?: string
}

/** Adder's keypair material — the device must already be in the current epoch. */
export interface AdderKeys {
  edPriv: string
  edPub: string
  kemPriv: string
  /** Adder's signing suite (governs the entry's `addedSig`). Absent ⇒ `ed25519`. */
  alg?: Alg
}

/** Optional knobs shared by the recipient-mutation helpers. */
export interface RecipientMutationOpts {
  /**
   * Ed25519 pubkeys (hex) of adders the caller trusts to have written keyring
   * entries — the same pin `createKeyringEncryptor` accepts. The `addedSig`
   * audit signature is *self-attesting* (any key signs its own entry), so a
   * hostile server can REPLACE the caller's own entry with one that wraps an
   * attacker-chosen CEK to the caller's KEM pubkey (every field is derivable
   * from the caller's *public* key) and self-sign it. `recoverCurrentCek`
   * would then unwrap that forged CEK and re-wrap it for the new recipient.
   * When `trustedAdders` is set, entries whose `addedBy` is not listed are
   * ignored, closing that substitution. Strongly recommended whenever the
   * server is not fully trusted.
   */
  trustedAdders?: string[]
}

/**
 * Resolve the mandatory `trustedAdders` pin into a Set, or throw. The mutation
 * helpers recover the current CEK from a server-supplied keyring; without a
 * provenance pin a hostile server could substitute a forged entry (the
 * `addedSig` is self-attesting). Fail closed rather than mutate off unverified
 * key material.
 */
function requireTrustedAdders(trustedAdders: string[] | undefined, fn: string): Set<string> {
  if (!trustedAdders || trustedAdders.length === 0) {
    throw new Error(
      `${fn}: \`trustedAdders\` is required — pass the Ed25519 pubkey(s) you trust to grant ` +
        `keyring access (e.g. the collection owner's root key). Without it a hostile server ` +
        `could substitute a wrapped-key entry (the addedSig is self-attesting).`,
    )
  }
  return new Set(trustedAdders)
}

function pullPathFor(collectionName: string): string {
  return `/pull/${keyringPathFor(collectionName)}`
}

function pushPathFor(collectionName: string): string {
  return `/push/${keyringPathFor(collectionName)}`
}

/**
 * Pulls the current keyring document. Returns `null` if no keyring exists yet
 * (HTTP 404). Any other error propagates.
 */
async function pullKeyring(
  client: StarfishClient,
  collectionName: string,
): Promise<{ keyring: Keyring; hash: string } | null> {
  try {
    const result = await client.pull(pullPathFor(collectionName))
    return { keyring: result.data as unknown as Keyring, hash: result.hash }
  } catch (err) {
    if (err instanceof StarfishHttpError && err.status === 404) return null
    throw err
  }
}

/**
 * Locates the adder's wrapped key entry in the current epoch and recovers the
 * CEK. Throws if the adder is not a member of the current epoch.
 *
 * Each candidate entry's `addedSig` is verified before the unwrap attempt:
 * a tampered audit signature (e.g. `addedBy` or `addedAt` mutated by an
 * intermediary) causes the entry to be skipped, so the client never trusts
 * unattested wrap material. A single corrupted entry does not prevent the
 * adder from recovering the CEK via another valid entry in the same epoch.
 */
async function recoverCurrentCek(
  keyring: Keyring,
  adderKemPriv: string,
  trustedAdders?: Set<string>,
): Promise<Uint8Array> {
  const epochKey = String(keyring.currentEpoch)
  const epoch = keyring.epochs[epochKey]
  if (!epoch) {
    throw new Error(`Epoch ${keyring.currentEpoch} not found in keyring`)
  }

  // A valid epoch has unique subKems (enforced on write by addRecipient).
  // Duplicates mean the keyring was tampered with — e.g. a hostile server
  // injected an entry wrapping an attacker-chosen CEK to this adder's key.
  // Fail closed rather than risk recovering and re-wrapping a forged CEK.
  const seenSubKems = new Set<string>()
  for (const e of epoch.wrappedKeys) {
    if (seenSubKems.has(e.subKem)) {
      throw new Error(
        `Keyring epoch ${keyring.currentEpoch} has duplicate entries for subKem=${e.subKem} (tampering)`,
      )
    }
    seenSubKems.add(e.subKem)
  }

  let lastErr: unknown
  for (const entry of epoch.wrappedKeys) {
    if (trustedAdders && !trustedAdders.has(entry.addedBy)) {
      console.warn(
        `[starfish:recipients] skipping entry subKem=${entry.subKem} in epoch ${keyring.currentEpoch}: addedBy ${entry.addedBy} is not a trusted adder`,
      )
      continue
    }
    const sigOk = await verifyEntrySignature(entry, keyring.currentEpoch)
    if (!sigOk) {
      console.warn(
        `[starfish:recipients] skipping entry subKem=${entry.subKem} in epoch ${keyring.currentEpoch}: addedSig verification failed`,
      )
      continue
    }
    try {
      return await unwrapFromEntry(entry, adderKemPriv)
    } catch (err) {
      lastErr = err
    }
  }
  throw new Error(
    `Adder has no usable entry in current epoch ${keyring.currentEpoch}`,
    lastErr instanceof Error ? { cause: lastErr } : undefined,
  )
}

/**
 * Adds a new recipient to the current epoch without rotating it. Pulls the
 * current keyring, unwraps the CEK using the adder's KEM private key, appends
 * a new entry signed by the adder, and pushes the updated keyring back using
 * the prior hash as `baseHash` for conflict detection.
 *
 * Throws if the keyring document does not exist yet.
 */
export async function addRecipient(
  client: StarfishClient,
  collectionName: string,
  recipient: RecipientRef,
  adder: AdderKeys,
  opts: RecipientMutationOpts = {},
): Promise<void> {
  const pulled = await pullKeyring(client, collectionName)
  if (!pulled) {
    throw new Error(
      `Cannot add recipient: no keyring exists at ${keyringPathFor(collectionName)}. ` +
        `Create the keyring first.`,
    )
  }

  const trustedAdders = requireTrustedAdders(opts.trustedAdders, "addRecipient")
  const currentCek = await recoverCurrentCek(pulled.keyring, adder.kemPriv, trustedAdders)
  const next = await keyringAddRecipient(
    pulled.keyring,
    { edPrivHex: adder.edPriv, edPubHex: adder.edPub, alg: adder.alg },
    currentCek,
    recipient.subKem,
    undefined,
    recipient.kemAlg,
  )

  await client.push(
    pushPathFor(collectionName),
    next as unknown as Record<string, unknown>,
    pulled.hash,
  )
}

/**
 * Rotates the epoch, dropping the named recipients. Pulls the current
 * keyring, mints a fresh CEK, wraps it for every retained recipient, and
 * pushes the updated keyring using the prior hash as `baseHash`.
 *
 * Throws if the keyring document does not exist yet.
 */
export async function removeRecipient(
  client: StarfishClient,
  collectionName: string,
  removeSubKems: string[],
  adder: AdderKeys,
  opts: RecipientMutationOpts = {},
): Promise<{ newEpoch: number }> {
  const pulled = await pullKeyring(client, collectionName)
  if (!pulled) {
    throw new Error(
      `Cannot remove recipient: no keyring exists at ${keyringPathFor(collectionName)}.`,
    )
  }

  const epochKey = String(pulled.keyring.currentEpoch)
  const epoch = pulled.keyring.epochs[epochKey]
  if (!epoch) {
    throw new Error(`Epoch ${pulled.keyring.currentEpoch} not found in keyring`)
  }

  // Entries written by an untrusted adder (e.g. a recipient a hostile server
  // injected) are not carried into the new epoch — the rotation would otherwise
  // re-wrap the fresh CEK for them. `trustedAdders` is mandatory (fail closed).
  const trustedAdders = requireTrustedAdders(opts.trustedAdders, "removeRecipient")
  const removeSet = new Set(removeSubKems)
  const retainedRecipients = epoch.wrappedKeys
    .filter((e) => !removeSet.has(e.subKem))
    .filter((e) => trustedAdders.has(e.addedBy))
    .map((e) => ({ subKemHex: e.subKem, kemAlg: e.kemAlg }))

  const { keyring: rotated } = await rotateEpoch(
    pulled.keyring,
    { edPrivHex: adder.edPriv, edPubHex: adder.edPub, alg: adder.alg },
    retainedRecipients,
  )

  await client.push(
    pushPathFor(collectionName),
    rotated as unknown as Record<string, unknown>,
    pulled.hash,
  )

  return { newEpoch: rotated.currentEpoch }
}

/** One recipient projected for listing. */
export interface ListedRecipient {
  subKem: string
  addedBy: string
  addedAt: number
}

/**
 * Lists recipients in the current epoch, filtered by provenance. Returns only
 * entries whose `addedBy` is in `trustedAdders` AND whose `addedSig` verifies —
 * mirroring `createKeyringEncryptor`, so a forged or server-substituted entry
 * never surfaces in a membership/admin view.
 *
 * `trustedAdders` is REQUIRED (fail-closed): the keyring is fetched from an
 * untrusted server and the per-entry `addedSig` is self-attesting, so without a
 * provenance pin a hostile server could spoof the listing. Returns
 * `{epoch: 0, recipients: []}` if no keyring document exists yet.
 */
export async function listRecipients(
  client: StarfishClient,
  collectionName: string,
  opts: { trustedAdders?: string[] } = {},
): Promise<{ epoch: number; recipients: ListedRecipient[] }> {
  const trusted = requireTrustedAdders(opts.trustedAdders, "listRecipients")
  const pulled = await pullKeyring(client, collectionName)
  if (!pulled) return { epoch: 0, recipients: [] }

  const epoch = pulled.keyring.epochs[String(pulled.keyring.currentEpoch)]
  if (!epoch) {
    return { epoch: pulled.keyring.currentEpoch, recipients: [] }
  }

  const recipients: ListedRecipient[] = []
  for (const e of epoch.wrappedKeys as WrappedKeyEntry[]) {
    if (!trusted.has(e.addedBy)) continue
    if (!(await verifyEntrySignature(e, pulled.keyring.currentEpoch))) continue
    recipients.push({ subKem: e.subKem, addedBy: e.addedBy, addedAt: e.addedAt })
  }
  return { epoch: pulled.keyring.currentEpoch, recipients }
}

/**
 * Returns the current epoch number of the keyring document. Returns 0 if no
 * keyring exists yet.
 */
export async function currentEpoch(
  client: StarfishClient,
  collectionName: string,
): Promise<number> {
  const pulled = await pullKeyring(client, collectionName)
  if (!pulled) return 0
  return pulled.keyring.currentEpoch
}
